source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,233 @@
+/*******************************************************************************
+
+    Contains MerkleTree that makes binary tree of transactions.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.MerkleTree;
+
+import agora.common.Data;
+import agora.common.Hash;
+import agora.common.Transaction;
+
+import std.algorithm;
+
+
+/// Ditto
+public struct MerkleTree
+{
+    /// The root of `MerkleTree`
+    public MerkleNode* root;
+
+    /***************************************************************************
+
+        Calculate the hash value of the `MerkleTree`
+
+        Return:
+            The `Hash` representing this instance
+
+    ***************************************************************************/
+
+    public Hash getHash ()
+        nothrow @nogc @trusted
+    {
+        if (this.root is null)
+            return Hash.init;
+        else
+            return this.root.getHash();
+    }
+}
+
+/// The `MerkleNode` has one transaction or two `MerkleNode`
+public struct MerkleNode
+{
+    /// The left branch's ptr of the tree.
+    public MerkleNode* left;
+
+    /// The right branch's ptr of the tree.
+    public MerkleNode* right;
+
+    /// The transaction's ptr
+    public Transaction* tx;
+
+    /// The hash of this `MerkleNode`
+    public Hash hash;
+
+    /***************************************************************************
+
+        Create
+
+        Params:
+            left = The left branch's ptr of the tree.
+            right = The right branch's ptr of the tree.
+            tx = The transaction's ptr
+
+    ***************************************************************************/
+
+    /// Ctor
+    this (MerkleNode* left, MerkleNode* right, Transaction* tx = null)
+    {
```

Comment:
You might want to add a contract that `left` and `right` are both non `null` OR that `tx` is null.

Reply:
@Geod24  Is this a `document` ?

Reply:
I'm sorry I don't understand the question.

I mean to add:
```d
assert((left is null) == (right is null), "`right` and `left` should be both `null` or both non-`null`");
assert((left is null) != (tx is null), "Either `tx` should be `null`, or there should be no child nodes");
```